gcc/rust/hir/rust-hir-dump.cc

```diff
@@ -17,678 +17,2457 @@
 // <http://www.gnu.org/licenses/>.
 
 #include "rust-hir-dump.h"
+#include "rust-abi.h"
+#include "rust-hir-item.h"
+#include "rust-hir-path.h"
+#include "rust-hir-type.h"
+#include "rust-hir.h"
+#include <string>
 
 namespace Rust {
 namespace HIR {
 
-Dump::Dump (std::ostream &stream) : stream (stream) {}
+// Dump Format for HIR
+//
+// SomeHIRNode [
+//    field: ...
+//    field: ...
+//    field: ...
+// ]
+//
+// When a field is a collection of other HIR object:
+//   field {
+//     SomeHIRNode [ ... ]
+//     SomeOtherHIRNode [ ... ]
+//   }
+//
+// If a field is an empty collection:
+//  field: empty
+//
+// If a field is optional and is currently not holding anything:
+//   field: none
 
 void
-Dump::go (HIR::Crate &crate)
+Dump::go (HIR::Crate &e)
+{
+  begin ("Crate");
+  do_inner_attrs (e);
+  do_mappings (e.get_mappings ());
+
+  visit_collection ("items", e.get_items ());
+  end ("Crate");
+}
+
+std::string
+convert_param_kind_to_str (MaybeNamedParam::ParamKind pk)
+{
+  switch (pk)
+    {
+    case MaybeNamedParam::ParamKind::UNNAMED:
+      return "UNNAMED";
+    case MaybeNamedParam::ParamKind::IDENTIFIER:
+      return "IDENTIFIER";
+    case MaybeNamedParam::ParamKind::WILDCARD:
+      return "WILDCARD";
+    }
+  gcc_unreachable ();
+}
+
+std::string
+convert_new_bind_type_to_str (UseTreeRebind::NewBindType nbt)
+{
+  switch (nbt)
+    {
+    case UseTreeRebind::NewBindType::NONE:
+      return "NONE";
+    case UseTreeRebind::NewBindType::IDENTIFIER:
+      return "IDENTIFIER";
+    case UseTreeRebind::NewBindType::WILDCARD:
+      return "WILDCARD";
+    }
+  gcc_unreachable ();
+}
+
+std::string
+convert_mut_to_str (Mutability mut)
```

Comment:
potentially you could move this over to util/rust-common.h i think and leave it as an inline helper

Reply:
I've implemented this change, and renamed the function at the same time to `enum_to_str`, a bit more explicit.